### PR TITLE
make a couple values lazy so they behave more like functions

### PR DESCRIPTION
**Pull request by dwightguth**
I was working on reimplementing pattern matching and Jost pointed out an issue with that code, which is that I was treating where bindings like they would only be evaluated when they were used, when in fact they get evaluated when the function that has the where binding gets called. This was leading to some issues in my code, but I realized also that it might be causing some performance degradations in the new unification algorithm as well, since that code was written in a similar style. And indeed, we were definitely making the mistake of using strict evaluation to evaluate some things that shouldn't need to get evaluated. So I made this PR. I have not tested the performance yet; want to make sure the tests pass first.

### Estimate:

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

**Comment by dwightguth** [comment]
This is ready for review, but don't merge it yet; we should wait to see performance statistics first.

**Comment by ehildenb** [comment]
This data isn't very good, because it spans a decent chunk of time, but this branch is not a regression against KEVM b9580ca4, which is a bit outdated:

```
 Test                                            | log-log time | b9580ca4-prove-haskell time | (log-log/b9580ca4-prove-haskell) time 
-------------------------------------------------+--------------+-----------------------------+---------------------------------------
 specs/mcd/flipper-addu48u48-fail-rough-spec.k   | 86.33        | 104.95                      | 0.8225821819914244                    
 specs/mcd/end-subuu-pass-spec.k                 | 123.05       | 149.17                      | 0.8248977676476504                    
 specs/functional/storageRoot-spec.k             | 30.94        | 36.08                       | 0.857538802660754                     
 specs/mcd/dai-adduu-fail-rough-spec.k           | 72.02        | 83.61                       | 0.8613802176773113                    
 specs/mcd/flipper-ttl-pass-spec.k               | 147.73       | 170.14                      | 0.8682849418126249                    
 specs/examples/solidity-code-spec.md            | 168.71       | 193.89                      | 0.8701325493836712                    
 specs/mcd/flipper-tau-pass-spec.k               | 150.68       | 172.89                      | 0.8715368153160971                    
 erc20/ds/totalSupply-spec.k                     | 108.65       | 123.28                      | 0.8813270603504219                    
 specs/mcd/vat-subui-fail-rough-spec.k           | 253.51       | 287.35                      | 0.8822342091526012                    
 erc20/ds/approve-failure-spec.k                 | 109.18       | 123.56                      | 0.8836192942699903                    
 specs/mcd/vat-addui-fail-rough-spec.k           | 253.81       | 286.12                      | 0.8870753529987417                    
 specs/mcd/vat-mului-pass-spec.k                 | 262.71       | 294.99                      | 0.8905725617817553                    
 specs/mcd/vat-subui-pass-rough-spec.k           | 355.15       | 397.7                       | 0.8930098063867237                    
 specs/mcd/dai-symbol-pass-spec.k                | 220.23       | 245.64                      | 0.8965559355153884                    
 specs/mcd/vat-subui-pass-spec.k                 | 360.44       | 401.85                      | 0.8969515988552942                    
 erc20/ds/balanceOf-spec.k                       | 184.71       | 205.36                      | 0.8994448772886637                    
 specs/mcd/cat-exhaustiveness-spec.k             | 151.38       | 167.28                      | 0.9049497847919655                    
 specs/mcd/flopper-cage-pass-spec.k              | 238.94       | 262.95                      | 0.9086898649933448                    
 erc20/hkg/totalSupply-spec.k                    | 43.47        | 47.72                       | 0.9109388097233864                    
 specs/mcd/dsvalue-read-pass-summarize-spec.k    | 142.56       | 156.17                      | 0.9128513799065122                    
 specs/benchmarks/functional-spec.k              | 41.01        | 44.87                       | 0.9139737018052151                    
 specs/mcd/vat-deny-diff-fail-rough-spec.k       | 439.59       | 480.73                      | 0.9144218168202525                    
 specs/functional/merkle-spec.k                  | 19.22        | 21.01                       | 0.9148024750118989                    
 erc20/hkg/balanceOf-spec.k                      | 98.18        | 107.21                      | 0.9157727823897026                    
 specs/functional/lemmas-spec.k                  | 24.3         | 26.5                        | 0.9169811320754717                    
 specs/mcd/cat-file-addr-pass-rough-spec.k       | 244.69       | 266.56                      | 0.9179546818727491                    
 erc20/ds/transferFrom-failure-1-d-spec.k        | 440.08       | 478.67                      | 0.9193807842563769                    
 specs/mcd/vow-flog-fail-rough-spec.k            | 387.74       | 421.45                      | 0.9200142365642425                    
 specs/bihu/forwardToHotWallet-failure-2-spec.k  | 290.43       | 315.46                      | 0.920655550624485                     
 erc20/ds/transferFrom-failure-2-c-spec.k        | 290.97       | 315.5                       | 0.9222503961965136                    
 specs/bihu/forwardToHotWallet-failure-1-spec.k  | 209.16       | 226.77                      | 0.9223442254266436                    
 specs/benchmarks/overflow00-nooverflow-spec.k   | 139.13       | 150.82                      | 0.9224903858904655                    
 specs/benchmarks/storagevar02-nooverflow-spec.k | 121.07       | 131.23                      | 0.9225786786557952                    
 erc20/ds/transfer-failure-1-c-spec.k            | 375.01       | 405.53                      | 0.9247404630976747                    
 erc20/ds/transfer-failure-2-b-spec.k            | 266.78       | 288.11                      | 0.9259657769601887                    
 specs/benchmarks/storagevar00-spec.k            | 99.44        | 107.39                      | 0.925970760778471                     
 specs/bihu/forwardToHotWallet-failure-4-spec.k  | 554.04       | 597.64                      | 0.9270463824375879                    
 erc20/hkg/transfer-failure-2-spec.k             | 183.96       | 198.38                      | 0.9273112208892026                    
 specs/benchmarks/encodepacked-keccak01-spec.k   | 191.08       | 206.01                      | 0.9275277899131111                    
 specs/benchmarks/bytes00-spec.k                 | 215.55       | 232.25                      | 0.9280947255113026                    
 specs/benchmarks/ecrecover00-siginvalid-spec.k  | 294.93       | 317.52                      | 0.9288548752834468                    
 specs/benchmarks/keccak00-spec.k                | 233.67       | 251.54                      | 0.9289576210543055                    
 specs/benchmarks/storagevar01-spec.k            | 110.4        | 118.79                      | 0.9293711591884839                    
 specs/mcd/vat-move-diff-rough-spec.k            | 790.76       | 850.53                      | 0.929726170740597                     
 specs/benchmarks/staticarray00-spec.k           | 189.83       | 204.15                      | 0.9298554984080334                    
 erc20/ds/allowance-spec.k                       | 283.59       | 304.91                      | 0.9300777278541207                    
 specs/mcd/vow-fess-fail-rough-spec.k            | 419.61       | 450.99                      | 0.9304197432315573                    
 erc20/ds/transfer-failure-2-a-spec.k            | 541.34       | 581.48                      | 0.9309692508770723                    
 specs/benchmarks/requires01-a0le0-spec.k        | 101.52       | 109.03                      | 0.9311198752636889                    
 specs/benchmarks/requires01-a0gt0-spec.k        | 138.19       | 148.12                      | 0.9329597623548473                    
 erc20/hkg/allowance-spec.k                      | 134.35       | 143.99                      | 0.9330509063129383                    
 erc20/hkg/approve-spec.k                        | 167.62       | 179.5                       | 0.9338161559888579                    
 specs/benchmarks/staticloop00-a0lt10-spec.k     | 244.14       | 261.41                      | 0.9339351975823418                    
 specs/benchmarks/encode-keccak00-spec.k         | 375.92       | 402.34                      | 0.9343341452502859                    
 specs/benchmarks/overflow00-overflow-spec.k     | 135.47       | 144.84                      | 0.9353079259872963                    
 specs/benchmarks/ecrecover00-sigvalid-spec.k    | 377.08       | 403.11                      | 0.9354270546500955                    
 specs/bihu/forwardToHotWallet-failure-3-spec.k  | 326.42       | 348.65                      | 0.9362397820163488                    
 erc20/ds/approve-success-spec.k                 | 509.12       | 542.88                      | 0.9378131447096965                    
 specs/benchmarks/address00-spec.k               | 128.22       | 136.07                      | 0.942309105607408                     
 specs/benchmarks/storagevar03-spec.k            | 86.53        | 91.82                       | 0.9423872794598127                    
 erc20/hkg/transfer-success-2-spec.k             | 229.1        | 243.09                      | 0.9424492986136821                    
 erc20/hkg/transfer-success-1-spec.k             | 380.45       | 402.34                      | 0.9455932793160015                    
 specs/benchmarks/storagevar02-overflow-spec.k   | 118.36       | 124.96                      | 0.9471830985915494                    
 specs/examples/sum-to-n-spec.k                  | 27.56        | 29.07                       | 0.9480564155486756                    
 specs/mcd/dsvalue-peek-pass-rough-spec.k        | 312.19       | 329.28                      | 0.948098882410107                     
 erc20/hkg/transfer-failure-1-spec.k             | 310.66       | 327.48                      | 0.9486380847685355                    
 specs/mcd/dsvalue-read-pass-spec.k              | 334.05       | 352.09                      | 0.9487631003436622                    
 specs/functional/infinite-gas-spec.k            | 21.39        | 22.47                       | 0.9519359145527371                    
 erc20/hkg/transferFrom-failure-2-spec.k         | 421.38       | 441.98                      | 0.9533915561790126                    
 erc20/hkg/transferFrom-success-2-spec.k         | 542.24       | 566.6                       | 0.9570067066713731                    
 specs/mcd/functional-spec.k                     | 25.41        | 26.55                       | 0.9570621468926553                    
 specs/opcodes/evm-optimizations-spec.md         | 21.21        | 22.1                        | 0.9597285067873303                    
 TOTAL                                           | 16428.34     | 17812.469999999998          | 0.9222943252676357  
```

**Review by ana-pantilie — APPROVED** [review]
LGTM, maybe we should wait for @jberthold to also take a look before merging.

**Review by jberthold — APPROVED** [review]
LGTM.
Surely we'll have more of these cases - I'd love to tryt just switching off `Strict` and measuring that. (we would certainly also find some leaks but could plug those in a targeted manner).

**Pull request merged**
